crates/ruma-client-api/src/knock/knock_room.rs

```diff
@@ -23,22 +23,129 @@ pub mod v3 {
     };
 
     /// Request type for the `knock_room` endpoint.
-    #[request(error = crate::Error)]
+    #[derive(Clone, Debug)]
+    #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
     pub struct Request {
         /// The room the user should knock on.
-        #[ruma_api(path)]
+        // #[ruma_api(path)]
         pub room_id_or_alias: OwnedRoomOrAliasId,
 
         /// The reason for joining a room.
-        #[serde(skip_serializing_if = "Option::is_none")]
+        // #[serde(skip_serializing_if = "Option::is_none")]
         pub reason: Option<String>,
 
         /// The servers to attempt to knock on the room through.
         ///
         /// One of the servers must be participating in the room.
-        #[ruma_api(query)]
+        // #[ruma_api(query)]
+        // #[serde(default, skip_serializing_if = "<[_]>::is_empty")]
+        pub via: Vec<OwnedServerName>,
+    }
+    /// Data in the request's query string.
+    #[cfg_attr(feature = "client", derive(serde::Serialize))]
+    #[cfg_attr(feature = "server", derive(serde::Deserialize))]
+    struct RequestQuery {
+        /// The servers to attempt to knock on the room through.
+        #[serde(default, skip_serializing_if = "<[_]>::is_empty")]
+        via: Vec<OwnedServerName>,
+
+        /// The servers to attempt to knock on the room through.
         #[serde(default, skip_serializing_if = "<[_]>::is_empty")]
-        pub server_name: Vec<OwnedServerName>,
+        #[deprecated = "Since Matrix 1.12, clients should use `knock::knock_room::v3::RequestQuery::via`."]
+        server_name: Vec<OwnedServerName>,
+    }
+
+    /// Data in the request's body.
+    #[cfg_attr(feature = "client", derive(serde::Serialize))]
+    #[cfg_attr(feature = "server", derive(serde::Deserialize))]
+    struct RequestBody {
+        /// The reason for joining a room.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub reason: Option<String>,
+    }
+
+    #[cfg(feature = "client")]
+    impl ruma_common::api::OutgoingRequest for Request {
+        type EndpointError = crate::Error;
+        type IncomingResponse = Response;
+
+        const METADATA: Metadata = METADATA;
+
+        #[allow(deprecated)]
+        fn try_into_http_request<T: Default + bytes::BufMut>(
+            self,
+            base_url: &str,
+            access_token: ruma_common::api::SendAccessToken<'_>,
+            considering_versions: &'_ [ruma_common::api::MatrixVersion],
+        ) -> Result<http::Request<T>, ruma_common::api::error::IntoHttpError> {
+            use http::header::{self, HeaderValue};
+
+            let query_string = serde_html_form::to_string(RequestQuery {
+                server_name: self.via.to_owned(),
```

Comment:
This is already an owned type so `.clone()` makes more sense.

```suggestion
                server_name: self.via.clone(),
```